producer/writer/writer.go

```diff
@@ -0,0 +1,258 @@
+		topicUpdateError:    scope.Counter("topic-update-error"),
+		invalidTopicUpdate:  scope.Counter("invalid-topic"),
+		invalidShard:        scope.Counter("invalid-shard"),
+		numConsumerServices: scope.Gauge("num-consumer-services"),
+	}
+}
+
+type writer struct {
+	sync.RWMutex
+
+	topic  string
+	ts     topic.Service
+	mPool  messagePool
+	opts   Options
+	logger log.Logger
+
+	value                  watch.Value
+	initType               initType
+	numShards              uint32
+	consumerServiceWriters map[string]consumerServiceWriter
+	filterRegistry         map[string]producer.FilterFunc
+	isClosed               bool
+	m                      writerMetrics
+
+	processFn watch.ProcessFn
+}
+
+// NewWriter creates a new writer.
+func NewWriter(opts Options) producer.Writer {
+	w := &writer{
+		topic:                  opts.TopicName(),
+		ts:                     opts.TopicService(),
+		mPool:                  newMessagePool(opts.MessagePoolOptions()),
+		opts:                   opts,
+		logger:                 opts.InstrumentOptions().Logger(),
+		initType:               failOnError,
+		consumerServiceWriters: make(map[string]consumerServiceWriter),
+		filterRegistry:         make(map[string]producer.FilterFunc),
+		isClosed:               false,
+		m:                      newWriterMetrics(opts.InstrumentOptions().MetricsScope()),
+	}
+	w.processFn = w.process
+	return w
+}
+
+func (w *writer) Write(d producer.RefCountedData) error {
+	w.RLock()
+	isClosed := w.isClosed
+	m := w.consumerServiceWriters
+	numShards := w.numShards
+	w.RUnlock()
+	if isClosed {
+		d.Drop()
+		return errWriterClosed
+	}
+	shard := d.Shard()
+	if shard >= numShards {
+		w.m.invalidShard.Inc(1)
+		d.Drop()
+		return fmt.Errorf("could not write data for shard %d which is larger than max shard id %d", shard, numShards-1)
+	}
+	// NB(cw): Need to inc ref here in case a consumer service
+	// writes the data too fast and close the data.
+	d.IncRef()
+	for _, csw := range m {
+		csw.Write(d)
+	}
+	d.DecRef()
+	return nil
+}
+
+func (w *writer) Init() error {
+	w.mPool.Init()
+	newUpdatableFn := func() (watch.Updatable, error) {
+		return w.ts.Watch(w.topic)
+	}
+	getUpdateFn := func(value watch.Updatable) (interface{}, error) {
+		t, err := value.(topic.Watch).Get()
+		if err != nil {
+			w.m.invalidTopicUpdate.Inc(1)
+			return nil, err
+		}
+		return t, nil
+	}
+	vOptions := watch.NewOptions().
+		SetInitWatchTimeout(w.opts.PlacementWatchInitTimeout()).
+		SetInstrumentOptions(w.opts.InstrumentOptions()).
+		SetNewUpdatableFn(newUpdatableFn).
+		SetGetUpdateFn(getUpdateFn).
+		SetProcessFn(w.processFn)
+	w.value = watch.NewValue(vOptions)
+	return w.value.Watch()
+}
+
+func (w *writer) process(update interface{}) error {
+	t := update.(topic.Topic)
+	// We don't allow changing number of shards for topics, it will be
+	// prevented on topic service side, but also being defensive here as well.
+	if w.numShards != 0 && w.numShards != t.NumberOfShards() {
```

Comment:
Why do we need to check it against 0?

Reply:
To make sure the numShards is initialized, when we create a writer, we don't know the value for numShards, it will be set after the first topic came in.

Reply:
Yeah makes sense.